main.go

```diff
@@ -0,0 +1,367 @@
+		return 0, 0, errors.New("missing end parameter")
+	}
+	startInt, err := strconv.ParseInt(start, 10, 64)
+	if err != nil || startInt < 0 {
+		return 0, 0, errors.New("invalid start parameter")
+	}
+	endInt, err := strconv.ParseInt(end, 10, 64)
+	if err != nil || endInt < 0 {
+		return 0, 0, errors.New("invalid end parameter")
+	}
+	if endInt <= startInt {
+		return 0, 0, errors.New("end must be greater than start")
+	}
+	return startInt, endInt, nil
+}
+
+// tile represents important numbers about a tile: where it starts, where it ends, its size,
+// and what CT backend URL it exists on (or is anticipated to exist on), and what s3 prefix
+// it should be stored/retrieved under.
+type tile struct {
+	start    int64
+	end      int64
+	size     int64
+	logURL   string
+	s3prefix string
+}
+
+// makeTile returns a tile of size `size` that contains the given `start` position.
+// The resulting tile's `start` will be equal to or less than the requested `start`.
+func makeTile(start, size int64, backend string, s3prefix string) tile {
+	tileOffset := start % size
+	tileStart := start - tileOffset
+	return tile{
+		start:    tileStart,
+		end:      tileStart + size,
+		size:     size,
+		logURL:   backend,
+		s3prefix: s3prefix,
+	}
+}
+
+// key returns the S3 key for the tile.
+func (t tile) key() string {
+	return fmt.Sprintf("%s/tile_size=%d/%d.cbor.gz", t.s3prefix, t.size, t.start)
+}
+
+// entries corresponds to the JSON response to the CT get-entries endpoint.
+// https://datatracker.ietf.org/doc/html/rfc6962#section-4.6
+//
+// It is marshaled and unmarshaled to/from JSON and CBOR.
+type entries struct {
+	Entries []entry `json:"entries"`
+}
+
+// entry corresponds to a single entry in the CT get-entries endpoint.
+//
+// Note: the JSON fields are base64. For fields of type `[]byte`, Go's encoding/json
+// automagically decodes base64.
+type entry struct {
+	LeafInput []byte `json:"leaf_input"`
+	ExtraData []byte `json:"extra_data"`
+}
+
+// statusCodeError indicates the backend returned a non-200 status code, and contains
+// the response body. This allows passing through that status code and body to the requester.
+type statusCodeError struct {
+	statusCode int
+	body       []byte
+}
+
+func (s statusCodeError) Error() string {
+	return fmt.Sprintf("backend responded with status code %d and body:\n%s", s.statusCode, string(s.body))
+}
+
+// getTileFromBackend fetches a tile of entries from the backend.
+//
+// If the backend returns a non-200 status code, it returns a statusCodeError,
+// so the caller can handle that case specially by propagating the backend's
+// status code (for instance, 400 or 404).
+func getTileFromBackend(ctx context.Context, path string, t tile) (*entries, error) {
+	url := fmt.Sprintf("%s/%s?start=%d&end=%d", t.logURL, path, t.start, t.end)
+	r, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create backend Request object: %s", err)
```

Comment:
```suggestion
		return nil, fmt.Errorf("unable to create backend Request object: %w", err)
```
more instances of this below, too